extensions/amp-geo/0.1/amp-geo.js

```diff
@@ -199,14 +202,97 @@ export class AmpGeo extends AMP.BaseElement {
       this.mode_ = mode.GEO_HOT_PATCH;
       this.country_ = trimmedCountryMatch[1];
     } else if (trimmedCountryMatch[0] === '' && !getMode(this.win).localDev) {
-      // We were not patched, if we're not in dev this is an error
-      // and we leave the country at the default 'unknown'
-      this.error_ = true;
-      dev().error(
-        TAG,
-        'GEONOTPATCHED: amp-geo served unpatched, ISO country not set'
-      );
+      // We were not patched, fall back to API if available
+      if (urls.geoApi) {
+        this.mode_ = mode.GEO_API;
+      } else {
+        // if we're not in dev this is an error
+        // and we leave the country at the default 'unknown'
+        this.error_ = true;
+        dev().error(
+          TAG,
+          'GEONOTPATCHED: amp-geo served unpatched, ISO country not set'
+        );
+      }
+    }
+
+    if (this.mode_ !== mode.GEO_API) {
+      return Promise.resolve();
+    }
+
+    return this.fetchCountry_().then(country => {
+      if (country) {
+        this.country_ = country;
+      } else {
+        // if API request fails, leave the country at the default 'unknown'
+        this.error_ = true;
+        dev().error(
+          TAG,
+          'GEONOTPATCHED: amp-geo served unpatched and API response not valid, ISO country not set'
+        );
+      }
+    });
+  }
+
+  /**
+   * Fetch country from API defined in config.urls
+   *
+   * JSON schema of Geo API response - version 0.1:
+   * {
+   *   "$schema": "http://json-schema.org/draft-07/schema#",
+   *   "type": "object",
+   *   "properties": {
+   *     "country": {
+   *       "type": "string",
+   *       "title": "ISO 3166-1 alpha-2 (case insensitive) country code of client request",
+   *       "default": "",
+   *       "pattern": "^[a-zA-Z]{2}$"
+   *     }
+   *   },
+   *   "required": [
+   *     "country"
+   *   ]
+   * }
+   *
+   * Sample response:
+   * {
+   *   "country": "de"
+   * }
+   *
+   * @return {Promise<(string|null)>}
+   */
+  fetchCountry_() {
+    if (typeof urls.geoApi !== 'string') {
```

Comment:
@jridgewell Did we require `https` in `urls`? If not we should use `assertHttpsUrl`

Reply:
HTTPS sounds like a good requirement to protect user privacy (physical location). `assertHttpsUrl` is geared more towards AMPHTML elements that have disallowed `src` attribute values. I opted for `isSecureUrlDeprecated`, which also makes allowances for localhost testing, since it is a bit more generic and less tied to `AmpDoc`.